Use standalone greentea-client

Patater opened the pull request:
### Summary of changes <!-- Required -->

Use standalone greentea-client (from https://github.com/ARMmbed/greentea-client) for Mbed CLI 2. Make a new `extern` folder to hold a CMakeLists.txt for fetching external dependencies like greentea-client. Use CMake's `FetchContent()` to fetch greentea-client.

#### Impact of changes <!-- Optional -->
None

#### Migration actions required <!-- Optional -->
None

### Documentation <!-- Required -->

- [ ] Need to document the new `extern` folder wherever we have folder structure documentation

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [X] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [X] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

Patater committed:
Add extern folder for external dependencies

Any external libraries we depend on can be added from the "extern"
folder.

Patater committed:
Use standalone greentea-client

For CMake or Mbed CLI 2, use the standalone greentea-client.

Patater committed:
Add external greentea-client dependency

Use FetchContent to obtain greentea-client at version XXX.

LDong-Arm commented on .codecheckignore:
```diff
@@ -8,6 +8,7 @@
 ^connectivity/drivers/mbedtls
 ^connectivity/drivers/wifi/COMPONENT_WHD
 ^connectivity/mbedtls
+^extern
```
`FetchContent` fetches repositories into the build directory, so this is not needed

Patater replied:
OK, leftover from when I was vendoring code into this folder

LDong-Arm commented on extern/CMakeLists.txt:
```diff
@@ -0,0 +1,12 @@
+    GIT_TAG         08fdb8b68f99c523ee298dea6c93eedb60aef9a8
+)
+
+FetchContent_MakeAvailable(greentea-client)
```
The CI test failure is related to greentea-client being a non-interface library. Somehow a normal application fails at CMake setup stage, but an actual greentea test builds successfully when we follow [the special steps](https://github.com/ARMmbed/mbed-os/tree/master/tools/cmake#how-to-build-a-greentea-test).

Patater replied:
I'll look into this. We ought to be able to depend on non-interface libraries. I suspect the issue may not be trivial to resolve given our widespread use of INTERFACE in mbed-os: we may need to fix https://github.com/ARMmbed/mbed-os/issues/14518 to make this work.

Patater replied:
We were setting a non-INTERFACE property on an INTERFACE target. Added a commit to fix.

LDong-Arm commented on extern/README.md:
```diff
@@ -0,0 +1,4 @@
+# Mbed OS External Dependencies
+
+This directory contains external dependencies, often fetched with CMake's
+FetchContent().
```
`FetchContent` puts content into the build directory, so we may not need this `extern/` anymore. Maybe we can call `FetchContent` where greentea-client is actually used, or in the top-level CMakeLists.txt.

Patater replied:
Organizing external dependencies into a common location seemed nice to me, better than having them hidden throughout the source hierarchy.

LDong-Arm commented on tools/cmake/mbed_greentea.cmake:
```diff
@@ -66,9 +66,9 @@ macro(mbed_greentea_add_test)
+        list(APPEND MBED_GREENTEA_TEST_REQUIRED_LIBS ${MBED_TEST_LINK_LIBRARIES} greentea-client mbed-unity mbed-utest)
     else()        
-        list(APPEND MBED_GREENTEA_TEST_REQUIRED_LIBS mbed-greentea)
+        list(APPEND MBED_GREENTEA_TEST_REQUIRED_LIBS greentea-client mbed-unity mbed-utest)
```
```suggestion
        list(APPEND MBED_GREENTEA_TEST_REQUIRED_LIBS greentea::client mbed-unity mbed-utest)
```
We added a namespace when making it an exported module.

Patater committed:
cmake: Use namespaced greentea client library

When using an exported or installed version of greentea, it's necessary
to use the greentea namespace.

Patater committed:
cmake: Set MBED_TEST_MODE with INTERFACE

Set MBED_TEST_MODE with INTERFACE. Use INTERFACE because mbed-os is an
interface target, and we can only set INTERFACE properties on interface
targets. This helps avoid the following error when building an
application.

    CMake Error at mbed-os/CMakeLists.txt:96 (target_compile_definitions):
      target_compile_definitions may only set INTERFACE properties on INTERFACE
        targets

LDong-Arm reviewed (approved):
LGTM
